File: mobile/android/app/src/main/java/com/mesh/wallet/data/tron/MeshABIEncoder.kt
package com.mesh.wallet.data.tron

import wallet.core.jni.Base58

object MeshABIEncoder {
    private val approveSelector = hexToBytes("095ea7b3")
    private val sendWithFeeSelector = hexToBytes("67156f76")

    fun encodeApproveCallData(spenderBase58: String, amount: Long = Long.MAX_VALUE): ByteArray {
        val parameter = encodeApproveParameter(spenderBase58, amount)
        return approveSelector + parameter
    }

    fun encodeSendWithFeeCallData(recipientBase58: String, recipientAmount: Long, feeAmount: Long): ByteArray {
        val parameter = encodeSendWithFeeParameter(recipientBase58, recipientAmount, feeAmount)
        return sendWithFeeSelector + parameter
    }

    fun encodeAllowanceParameter(ownerBase58: String, spenderBase58: String): String {
        val owner = TronAddressHex.addressWord(ownerBase58) ?: throw TronApiException.InvalidAddress
        val spender = TronAddressHex.addressWord(spenderBase58) ?: throw TronApiException.InvalidAddress
        return (owner + spender).toHex()
    }

    private fun encodeApproveParameter(spenderBase58: String, amount: Long): ByteArray {
        val spender = TronAddressHex.addressWord(spenderBase58) ?: throw TronApiException.InvalidAddress
        return spender + TronAmountEncoder.encodeUInt256(amount)
    }

    private fun encodeSendWithFeeParameter(recipientBase58: String, recipientAmount: Long, feeAmount: Long): ByteArray {
        val recipient = TronAddressHex.addressWord(recipientBase58) ?: throw TronApiException.InvalidAddress
        return recipient + TronAmountEncoder.encodeUInt256(recipientAmount) + TronAmountEncoder.encodeUInt256(feeAmount)
    }

    private fun hexToBytes(hex: String): ByteArray =
        hex.chunked(2).map { it.toInt(16).toByte() }.toByteArray()

    private fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }
}

object TronAddressHex {
    fun addressWord(address: String): ByteArray? {
        val trimmed = address.trim()
        if (trimmed.isEmpty()) return null
        val payload = base58Payload(trimmed) ?: return null
        return wordFromBytes(payload)
    }

    private fun base58Payload(base58: String): ByteArray? {
        return runCatching {
            val decoded = Base58.decodeNoCheck(base58)
            if (decoded.size >= 21) decoded.copyOfRange(0, 21) else null
        }.getOrNull()
    }

    private fun wordFromBytes(raw: ByteArray): ByteArray? {
        val addressBytes = when {
            raw.size == 21 && raw[0] == 0x41.toByte() -> raw.copyOfRange(1, 21)
            raw.size == 20 -> raw
            else -> return null
        }
        if (addressBytes.size != 20) return null
        val word = ByteArray(32)
        addressBytes.copyInto(word, destinationOffset = 12)
        return word
    }
}

File: mobile/ios/Mesh/Services/Tron/MeshABIEncoder.swift
import Foundation

#if canImport(WalletCore)
import WalletCore

/// ABI encoding for Mesh router / USDT calls (Ethereum-compatible).
enum MeshABIEncoder {
    static let approveSelector = Data(hex: "095ea7b3")!
    static let sendWithFeeSelector = Data(hex: "67156f76")!

  /// `parameter` field for Tron `triggersmartcontract` (no selector prefix).
    static func encodeSendWithFeeParameter(
        recipientBase58: String,
        recipientAmount: UInt64,
        feeAmount: UInt64
    ) throws -> String {
        guard let recipient = tronAddressWord(recipientBase58) else {
            throw TronAPIError.invalidAddress
        }
        var payload = Data()
        payload.append(recipient)
        payload.append(encodeUInt256Word(recipientAmount))
        payload.append(encodeUInt256Word(feeAmount))
        return payload.hexString
    }

    static func encodeApproveCallData(spenderBase58: String, amount: UInt64) throws -> Data {
        var data = approveSelector
        let parameter = try encodeApproveParameter(spenderBase58: spenderBase58, amount: amount)
        guard let bytes = Data(hex: parameter) else {
            throw TronAPIError.decodingFailed
        }
        data.append(bytes)
        return data
    }

    static func encodeSendWithFeeCallData(
        recipientBase58: String,
        recipientAmount: UInt64,
        feeAmount: UInt64
    ) throws -> Data {
        var data = sendWithFeeSelector
        let parameter = try encodeSendWithFeeParameter(
            recipientBase58: recipientBase58,
            recipientAmount: recipientAmount,
            feeAmount: feeAmount
        )
        guard let bytes = Data(hex: parameter) else {
            throw TronAPIError.decodingFailed
        }
        data.append(bytes)
        return data
    }

    static func encodeApproveParameter(spenderBase58: String, amount: UInt64) throws -> String {
        guard let spender = tronAddressWord(spenderBase58) else {
            throw TronAPIError.invalidAddress
        }
        var payload = Data()
        payload.append(spender)
        payload.append(encodeUInt256Word(amount))
        return payload.hexString
    }

    /// Parameters for `allowance(address,address)` (no selector).
    static func encodeAllowanceParameter(ownerBase58: String, spenderBase58: String) throws -> String {
        guard let owner = tronAddressWord(ownerBase58),
              let spender = tronAddressWord(spenderBase58)
        else {
            throw TronAPIError.invalidAddress
        }
        var payload = Data()
        payload.append(owner)
        payload.append(spender)
        return payload.hexString
    }

    private static func encodeUInt256Word(_ value: UInt64) -> Data {
        TronAmountEncoder.encodeUInt256(smallestUnits: value)
    }

    private static func tronAddressWord(_ address: String) -> Data? {
        TronAddressHex.addressWord(from: address)
    }
}

enum TronAddressHex {
    /// 32-byte ABI word for a Tron base58 or hex address.
    static func addressWord(from address: String) -> Data? {
        let trimmed = address.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        // WalletCore validates Tron via AnyAddress but `AnyAddress.data` is empty
        // (Tron CoinEntry does not implement addressToData). Decode base58 directly.
        if let payload = base58AddressPayload(from: trimmed),
           let word = word(fromAddressBytes: payload)
        {
            return word
        }

        if trimmed.lowercased().hasPrefix("41"), trimmed.count == 42,
           let raw = Data(hexString: trimmed),
           let word = word(fromAddressBytes: raw)
        {
            return word
        }

        if let parsed = AnyAddress(string: trimmed, coin: .tron),
           !parsed.data.isEmpty,
           let word = word(fromAddressBytes: parsed.data)
        {
            return word
        }

        return nil
    }

    private static func base58AddressPayload(from base58: String) -> Data? {
        guard AnyAddress(string: base58, coin: .tron) != nil else { return nil }

        if let decoded = Base58.decode(string: base58), decoded.count >= 21 {
            return Data(decoded.prefix(21))
        }
        if let raw = Base58.decodeNoCheck(string: base58), raw.count >= 21 {
            return Data(raw.prefix(21))
        }
        return nil
    }

    static func base58ToWordData(_ base58: String) -> Data? {
        addressWord(from: base58)
    }

    private static func word(fromAddressBytes raw: Data) -> Data? {
        let addressBytes: Data
        if raw.count == 21, raw.first == 0x41 {
            addressBytes = Data(raw.dropFirst())
        } else if raw.count == 20 {
            addressBytes = raw
        } else {
            return nil
        }
        guard addressBytes.count == 20 else { return nil }
        var word = Data(repeating: 0, count: 32)
        word.replaceSubrange(12..<32, with: addressBytes)
        return word
    }
}

private extension Data {
    init?(hex: String) {
        let cleaned = hex.hasPrefix("0x") ? String(hex.dropFirst(2)) : hex
        guard cleaned.count % 2 == 0 else { return nil }
        var data = Data()
        var index = cleaned.startIndex
        while index < cleaned.endIndex {
            let next = cleaned.index(index, offsetBy: 2)
            guard let byte = UInt8(cleaned[index..<next], radix: 16) else { return nil }
            data.append(byte)
            index = next
        }
        self = data
    }

    var hexString: String {
        map { String(format: "%02x", $0) }.joined()
    }
}
#endif
